Add portus chart

rendhalver opened the pull request:
This is a resubmit of #1284.
Thanks to @so0k for all the hard work!
This is currently the same as the last commit in that PR.

It looks like there are a few differences in layout since that PR was submitted so I will try to address those concerns as I go.
There were also a few concerns not address in the original PR so I will address those as well.

Commit:
Add portus chart

Commit:
Fix formating in portus/values.yaml

Commit:
Fix formatting in portus/vlaues.yaml

unguiculus commented:
Please check our best practices, especially those on naming resources and standard labels. Resource metadata should generally look like so:

```yaml
name: {{ template "portus.fullname" . }}
labels:
  heritage: {{ .Release.Service }}
  release: {{ .Release.Name }}
  chart: {{ .Chart.Name }}-{{ .Chart.Version }}
  app: {{ template "portus.name" . }}
```

rendhalver commented:
Thanks @unguiculus I will get those fixed.

rendhalver commented:
@unguiculus This Chart has multiple components but there isn't a definition $component.name in the _helpers.tpl file and I am not sure of the proper way to do that.
Can you point me to some docs on how to achieve that?

On a side note I am very tempted to split out the registry component into it's own chart and then add that as a requirement for this chart so the templates are a bit simpler.

I am also new to writing Charts so I am still learning how best to do them.

unguiculus commented:
There's a chart for Docker registry already: https://github.com/kubernetes/charts/tree/master/incubator/docker-registry

rendhalver commented:
True. I forgot about that Chart.
I had already discounted that because doesn't have the options we need to setup the registry for Portus.
Maybe I can expand that one so it can be configured with those extra options.

As for my other question what is the best way to define those $component.name variables?

Commit:
Fix app label in templates

rendhalver commented:
@unguiculus I think worked out what I needed to change.
The latest commit addresses the label inconsistencies.

unguiculus commented:
Please create a PR for the registry chart to add the stuff you need. We should also see that we can move it to stable so you can use it.

rendhalver commented:
Will do.
Thanks for the help @unguiculus It is much appreciated. :)

so0k commented:
The reason I had them bundled was the auth token service configuration required on the registry side.

If the registry chart exposes that, it should be fine to define it as a dependency though

rendhalver commented:
@unguiculus What else needs changing in the docker-registry chart to move it to stable?

unguiculus commented on stable/portus/Chart.yaml:
```diff
@@ -0,0 +1,16 @@
+apiVersion: v1
+name: portus
+home: http://port.us.org/
+version: 0.1.0
```
Add `appVersion`

unguiculus commented on stable/portus/Chart.yaml:
```diff
@@ -0,0 +1,16 @@
+  - https://github.com/docker/distribution/
+  - https://github.com/Nordstrom/kube-registry
+maintainers:
+  - name: Vincent De Smet
```
Use Github username

so0k replied:
can remove me if this is holding back the merge :)

Commit:
Switch to using docker-registry chart

Don't build a registry container in this chart.
Instead we will use the existing docker-registry Chart.

Commit:
Add support for LDAP authentication

Commit:
Add option to override ingress setup

viglesiasce commented:
/ok-to-test

rendhalver commented:
@viglesiasce This is dependent on my other PR for getting docker-registry usable and put of incubator #2801. Once that is sorted I can update this to reflect the new location of docker-registry.

Commit:
Setup secret for ldap bind password

unguiculus commented:
I think the Docker registry chart looks pretty good. I'd suggest you create a PR to move it to stable.

rendhalver commented:
It needs to be redone or rebased. Redoing it seems quicker at this stage. #2612 got merged that changed a lot of the docker-registry chart.
I will get that sorted when I get a chance.

rendhalver commented:
The docker-registry PR has been redone in #2926.

Commit:
Update docker-registry config options

Fix README

Commit:
Add my details to maintainers and add appVersion

rendhalver commented:
@unguiculus This is my PR to promote docker-registry to stable: #2944
I can update the requirements on this once that is merged.

Commit:
Add option to disable post delete job

Commit:
Update docker-registry and mariadb dependencies

Commit:
Disable minio by default

The chart fails to install with it set.
The default storage is set to filesystem

rendhalver commented:
/retest

unguiculus commented:
Let's try it again. Timeouts have been increased.

unguiculus commented:
/retest

rendhalver commented:
I ran a test locally and it's failing with default settings because it expects a tls secret setup with the cert for docker-registry.
It's listed in the README.

unguiculus commented:
You'll have to change it such that CI passes. Otherwise it can't be merged.

bonhamjp commented:
Hello @rendhalver the work on this chart is looking very good! 

I wasn't aware that anyone was working on implementing a Portus chart, until I came across this. I work at SUSE, and we have also been developing a Portus chart. Our work can be found at https://github.com/kubic-project/caasp-services/tree/master/contrib/helm-charts/portus

Our goal was to propose our chart here when we finished developing it. Would you be interested in collaborating?

rendhalver commented:
Hi @bonhamjp I wasn't aware SUSE was working on a Chart for Portus so I guess we both learned something. :)
Collaboration sounds good to me.

Are you using your Chart in production?

The portus parts of your chart look similar to mine and has some good features that would help get this one finished like generating a tls cert in the container.

I have tried to make this chart use other charts to deploy the components (docker-registry for example).
I am in k8s Slack as rendhalver as well if you want to chat about it there.

so0k commented:
@bonhamjp - I pinged portus maintainers on the portus issues related to k8s deployments almost a year ago (when I submitted the first chart proposal - mostly based on the work from Nordstorm)

However, I've noticed Portus 2.2 release branche was broken for several months and overall development seems stuck - which is why we no longer run Portus at Honestbee

Happy to see work is still ongoing

rendhalver commented:
@so0k I couldn't get 2.2 working in our setup either.
I will downgrade it in this chart till it does work.

so0k commented:
you're supposed to use an image tagged with a conference name, main reason I dropped Portus

rendhalver commented:
@so0k What do you mean by "image tagged with a conference name"?

so0k commented:
https://github.com/SUSE/Portus/issues/1418#issuecomment-355291208

rendhalver commented:
Thanks @so0k. I will keep an eve on that.

Commit:
Update to latest version of docker-registry

Commit:
Add values to configure portus registry settings

Commit:
Fix value names for registry settings

Commit:
Restart Portus deployemnt on configmap changes

Commit:
Add pod annotations

Move update strategy into values so it is customisable.
Update docker-registry version

Commit:
Add OWNERS file for stable/portus

Commit:
Add release label to Deployement and pods

Use that release label in the service so we don't include pods from other releeases.
Add annotations for the servive if set.
Add type to service with default of ClusterIP
Add ability to set NodePort if needed

Commit:
Update requirements.lok for portus

Commit:
Update docker-registry version

Commit:
Update requirements lock file

Commit:
Update pod labels to match new standards

Commit:
Fix github username for Vincent

rendhalver commented:
@unguiculus I just rebased this to pickup the new tests and It looks like we have test success now.
Do you have any further concerns about this one?

rendhalver commented:
I just realised this will need changes to work with Portus 2.3.
I will get those added as soon as I can.

Commit:
Add support for Portus 2.3

Some env vars are different in 2.3.
This changes the deploy to use those when we are running 2.3
It also adds support for the portus background jobs that were added in 2.3
Those jobs are run in a separate container in the pod

Commit:
Fix volumeMounts for portus background container

unguiculus commented on stable/portus/templates/portus-post-delete-job.yaml:
```diff
@@ -0,0 +1,38 @@
+    metadata:
+      name: "{{ .Release.Name }}"
+      labels:
+        app: {{ template "portus.name" . }}
```
Add `release` label to selector.

rendhalver replied:
Thanks for the reminder.
I will get that fixed.

unguiculus commented on stable/portus/templates/minio-create-bucket-job.yaml:
```diff
@@ -0,0 +1,35 @@
+      name: "{{ .Release.Name }}"
+      labels:
+        app: "{{ template "portus.name" . }}"
+        component: minio
```
Add `release` label to selector.

rendhalver replied:
Thanks for the reminder.
I will get that fixed.

Commit:
Add release to selector labels

Commit:
Generate TLS certs if not provided

Previosly we assumed the cert for the secrets existed.
This moves the creation of that secret into the chart.
If no certs are provided we generate them

Commit:
Remove log_level env var from background container

unguiculus committed:
Merge remote-tracking branch 'upstream/master' into HEAD

giacope commented on stable/portus/.helmignore:
```diff
@@ -0,0 +1,21 @@
+# Various IDEs
+.project
+.idea/
+*.tmproj
```
Please add the OWNERS file in the .helmignore - https://github.com/helm/charts#owning-and-maintaining-a-chart
